File: apps/api/plane/db/models/custom_field.py
# Copyright (c) 2023-present Plane Software, Inc. and contributors
# SPDX-License-Identifier: AGPL-3.0-only
# See the LICENSE file for details.

from django.db import models
from django.db.models import Q

from .project import ProjectBaseModel


class ProjectCustomField(ProjectBaseModel):
    FIELD_TYPE_CHOICES = (
        ("text", "Text"),
        ("number", "Number"),
        ("date", "Date"),
        ("boolean", "Boolean"),
        ("select", "Select"),
        ("multi_select", "Multi Select"),
    )

    name = models.CharField(max_length=255)
    field_type = models.CharField(max_length=40, choices=FIELD_TYPE_CHOICES, default="text")
    config = models.JSONField(default=dict, blank=True)
    sort_order = models.FloatField(default=65535)

    class Meta:
        verbose_name = "Project Custom Field"
        verbose_name_plural = "Project Custom Fields"
        db_table = "project_custom_fields"
        ordering = ("sort_order", "created_at")
        constraints = [
            models.UniqueConstraint(
                fields=["project", "name"],
                condition=Q(deleted_at__isnull=True),
                name="project_custom_field_unique_project_name_when_deleted_at_null",
            )
        ]

    def __str__(self):
        return f"{self.name} <{self.project.name}>"

File: packages/types/src/project/custom-field.ts
/**
 * Copyright (c) 2023-present Plane Software, Inc. and contributors
 * SPDX-License-Identifier: AGPL-3.0-only
 * See the LICENSE file for details.
 */

export type TCustomFieldType = "text" | "number" | "date" | "boolean" | "select" | "multi_select";

export interface IProjectCustomField {
  id: string;
  project_id: string;
  workspace_id: string;
  name: string;
  field_type: TCustomFieldType;
  config: {
    options?: string[];
    [key: string]: unknown;
  };
  sort_order: number;
  created_at: string;
  updated_at: string;
  created_by: string | null;
  updated_by: string | null;
}
